design/restic-backup-and-restore-progress.md

```diff
@@ -0,0 +1,114 @@
+```
+
+The [command factory for backup](https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/pkg/restic/command_factory.go#L37) will be updated to include the `--json` flag.
+The code to run the `restic backup` command (https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/pkg/controller/pod_volume_backup_controller.go#L241) will be changed to include a Goroutine that reads from the command's stdout stream.
+The implementation of this will largely follow [@jmontleon's PoC](https://github.com/fusor/velero/pull/4/files) of this.
+The Goroutine will periodically read the stream (every 10 seconds) and get the last printed status line.
+The line will be decoded as JSON and the `bytes_done` and `total_bytes` property will be read and the PodVolumeBackup will be patched to update `status.Progress.BytesDone` and `status.Progress.TotalBytes` respectively.
+
+Once the backup has completed successfully, the PodVolumeBackup will be patched to set `status.Progress.BytesDone = status.Progress.TotalBytes`.
+This is done since the main thread may cause early termination of the Goroutine once the operation has finished, preventing a final update to the `BytesDone` property.
+
+### restic restore progress
+
+The `restic stats <snapshot_id> --json` command provides information about the size of backups:
+
+```
+{"total_size":10558111744,"total_file_count":11}
+```
+
+Before beginning the restore operation, we can use the output of `restic stats` to get the total size of the backup.
+The PodVolumeRestore will be patched to set `status.Progress.TotalBytes` to the total size of the backup.
+
+The code to run the `restic restore` command will be changed to include a Goroutine that periodically (every 10 seconds) gets the current size of the volume.
+To get the current size of the volume, we will recursively walkthrough all files in the volume to accumulate the total size.
+The current total size is the number of bytes transferred so far and the PodVolumeRestore will be patched to update `status.Progress.BytesDone`.
+
+Once the restore has completed successfully, the PodVolumeRestore will be patched to set `status.Progress.BytesDone = status.Progress.TotalBytes`.
+This is done since the main thread may cause early termination of the Goroutine once the operation has finished, preventing a final update to the `BytesDone` property.
+
+### Velero CLI changes
+
+The output that describes detailed information about [PodVolumeBackups](https://github.com/heptio/velero/blob/559d62a2ec99f7a522924348fc4a173a0699813a/pkg/cmd/util/output/backup_describer.go#L349) and [PodVolumeRestores](https://github.com/heptio/velero/blob/559d62a2ec99f7a522924348fc4a173a0699813a/pkg/cmd/util/output/restore_describer.go#L160) will be updated to calculate and display a completion percentage from `status.Progress.TotalBytes` and `status.Progress.BytesDone` if available.
```

Comment:
Do you intend to add any changes to the `velero backup create --watch` functionality? Currently, this just prints `.`s until the backup completes.

If this is out of scope for this PR, that's fine, just wondering if you'd thought of it.

Reply:
Ah I hadn't thought of that, good point. I think it's a great idea, but perhaps out-of-scope for this issue. Let's create another issue for it and we can follow up with that later?

Reply:
Yep, seems like a separate issue.